Highlights are not repainted if associated live ranges are changed.

megangardner opened the pull request:
#### 306860296d33ebc428b319f6d705faa3a51f6e33
<pre>
Highlights are not repainted if associated live ranges are changed.
<a href="https://bugs.webkit.org/show_bug.cgi?id=262168">https://bugs.webkit.org/show_bug.cgi?id=262168</a>
rdar://116108148

Reviewed by Wenson Hsieh.

When a range is changed, we need to repaint the previous and the new range.
Fortunately, we already had a mechanism to get Positions from the ranges
so that we have the information we need at paint time. We can use these
stale positions to know what the range was before updating, and repaint the old
range, as well as the new range.
Also a rename for more clarity around setting flags for highlights in Range.

* LayoutTests/fast/repaint/highlight-paint-after-range-change-expected.txt: Added.
* LayoutTests/fast/repaint/highlight-paint-after-range-change.html: Added.
* Source/WebCore/Modules/highlight/Highlight.h:
(WebCore::HighlightRange::HighlightRange):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::collectHighlightRangesFromRegister):
(WebCore::repaintRange):
(WebCore::Document::updateHighlightPositions):
(WebCore::Document::collectRangeDataFromRegister): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::updateAssociatedHighlight):
(WebCore::Range::setStart):
(WebCore::Range::setEnd):
(WebCore::Range::nodeChildrenChanged):
(WebCore::Range::nodeChildrenWillBeRemoved):
(WebCore::Range::nodeWillBeRemoved):
(WebCore::Range::textInserted):
(WebCore::Range::textRemoved):
(WebCore::Range::textNodesMerged):
(WebCore::Range::textNodeSplit):
* Source/WebCore/dom/Range.h:

Canonical link: <a href="https://commits.webkit.org/268660@main">https://commits.webkit.org/268660@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/95c763d2f16729eed6eda1a777cce749ead5a5e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20276 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20709 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21349 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22177 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18924 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20508 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23966 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20879 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20359 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20400 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17633 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23027 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17569 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18436 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24691 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18645 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18612 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22664 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19193 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16299 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18398 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18272 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4884 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22739 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19018 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/repaint/highlight-paint-after-range-change-expected.txt:
```diff
@@ -0,0 +1,5 @@
+Highlight this first and then this.
```
Would be good to add an indication of how to read the test results. There's been cases where rebaselining these ended up in regressing the test.

Something like: "Test passes if [...]"

whsieh replied:
I think the `title` element in the test has a pretty reasonable description, but I suppose it could be moved into the body instead using either `js-test.js` + `description(…)`, or as a separate element (such that it shows up in the `-expected.txt`).

nt1m replied:
I think a description of which repaint rects (numbers) are expected is generally useful. It's possible for someone to rebaseline such a test with the wrong repaints happening. It has happened to me, and I've seen rebaselines of dumps where the rebaseline actually turned out to be wrong.

smfr commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -899,7 +907,7 @@ void Range::nodeChildrenChanged(ContainerNode& container)
     boundaryNodeChildrenChanged(m_start, container);
     boundaryNodeChildrenChanged(m_end, container);
-    m_didChangeHighlight = true;
+    m_didChangeForHighlight = true;
```
Should this only be changed if `m_isAssociatedWithHighlight` is true or does it not really matter?

megangardner replied:
it doesn't really matter, and I don't think checking bool to change another bool is faster than just changing it all the time.

smfr commented on LayoutTests/fast/repaint/highlight-paint-after-range-change.html:
```diff
@@ -0,0 +1,53 @@
+        }
+        if (window.testRunner)
+            testRunner.notifyDone();
+     }, 100);
```
You need to figure out how to avoid the 100ms timeout. Maybe it will just work now, or maybe you'll need an `await UIHelper.renderingUpdate()`

whsieh commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3056,8 +3057,20 @@ void Document::collectRangeDataFromRegister(Vector<WeakPtr<HighlightRange>>& hig
         for (auto& highlightRange : highlight.value->highlightRanges()) {
-            if (auto* liveRange = dynamicDowncast<Range>(highlightRange->range()); liveRange && liveRange->didChangeHighlight())
-                liveRange->resetDidChangeHighlight();
+            if (auto* liveRange = dynamicDowncast<Range>(highlightRange->range()); liveRange && liveRange->didChangeForHighlight())
```
Nit - `RefPtr` instead of raw pointers here?

whsieh commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3056,8 +3057,20 @@ void Document::collectRangeDataFromRegister(Vector<WeakPtr<HighlightRange>>& hig
+    auto sortedRange = makeSimpleRange(range);
+    if (is_gt(treeOrder<ComposedTree>(sortedRange.start, sortedRange.end)))
+        std::swap(sortedRange.start, sortedRange.end);
+    for (auto& node : intersectingNodes(sortedRange)) {
```
Can we use `Ref` for the node here?

whsieh reviewed (approved):
r=me, pending ntim's [earlier comment](https://github.com/WebKit/WebKit/pull/18266/files#r1338463207) as well.

megangardner committed:
Highlights are not repainted if associated live ranges are changed.
https://bugs.webkit.org/show_bug.cgi?id=262168
rdar://116108148

Reviewed by Wenson Hsieh.

When a range is changed, we need to repaint the previous and the new range.
Fortunately, we already had a mechanism to get Positions from the ranges
so that we have the information we need at paint time. We can use these
stale positions to know what the range was before updating, and repaint the old
range, as well as the new range.
Also a rename for more clarity around setting flags for highlights in Range.

* LayoutTests/fast/repaint/highlight-paint-after-range-change-expected.txt: Added.
* LayoutTests/fast/repaint/highlight-paint-after-range-change.html: Added.
* Source/WebCore/Modules/highlight/Highlight.h:
(WebCore::HighlightRange::HighlightRange):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::collectHighlightRangesFromRegister):
(WebCore::repaintRange):
(WebCore::Document::updateHighlightPositions):
(WebCore::Document::collectRangeDataFromRegister): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::updateAssociatedHighlight):
(WebCore::Range::setStart):
(WebCore::Range::setEnd):
(WebCore::Range::nodeChildrenChanged):
(WebCore::Range::nodeChildrenWillBeRemoved):
(WebCore::Range::nodeWillBeRemoved):
(WebCore::Range::textInserted):
(WebCore::Range::textRemoved):
(WebCore::Range::textNodesMerged):
(WebCore::Range::textNodeSplit):
* Source/WebCore/dom/Range.h:

Canonical link: https://commits.webkit.org/268660@main
